From 9067ce91f758050a0a0ccf0dcd917dde4090e56c Mon Sep 17 00:00:00 2001
From: Paul Thomson <paulthomson@google.com>
Date: Fri, 17 Apr 2026 10:07:56 +0100
Subject: [PATCH 3/9] Fix package annotations

...while we wait for the upstream fix to become available.

The upstream fix might be here:
https://github.com/JetBrains/intellij-community/commit/d64f98cda7d27d3938ea9826395c0c0e21637900

Bug: https://youtrack.jetbrains.com/issue/IDEA-383732
Bug: 478260770
Test: AnnotationHandlerTest
---
 .../src/com/intellij/psi/impl/file/PsiPackageImpl.java    | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/java/java-psi-impl/src/com/intellij/psi/impl/file/PsiPackageImpl.java b/java/java-psi-impl/src/com/intellij/psi/impl/file/PsiPackageImpl.java
index aa8d14da59d2..560cb51d9658 100644
--- a/java/java-psi-impl/src/com/intellij/psi/impl/file/PsiPackageImpl.java
+++ b/java/java-psi-impl/src/com/intellij/psi/impl/file/PsiPackageImpl.java
@@ -441,7 +441,13 @@ public class PsiPackageImpl extends PsiPackageBase implements PsiPackage, Querya
       }
 
       for (PsiClass aClass : getFacade().findClasses(getQualifiedName() + ".package-info", allScope())) {
-        ContainerUtil.addIfNotNull(modifiers, aClass.getModifierList());
+        PsiFile file = aClass.getContainingFile();
+        if (file instanceof PsiJavaFile) {
+          PsiPackageStatement stmt = ((PsiJavaFile)file).getPackageStatement();
+          if (stmt != null) {
+            ContainerUtil.addIfNotNull(modifiers, stmt.getAnnotationList());
+          }
+        }
       }
 
       PsiCompositeModifierList result = modifiers.isEmpty() ? null : new PsiCompositeModifierList(getManager(), modifiers);
-- 
2.54.0.rc1.513.gad8abe7a5a-goog

